### PR TITLE
Add the first version of workflow design

### PR DESCRIPTION
**Pull request by ywskycn**

### CODE REVIEW

File: doc/design/design_workflow.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+Tangent does autodiff by parsing the Python source code into an abstract syntax tree (AST), adding the backward steps into the AST tree, and printing the AST tree into a new snippet of Python code.
+
+We like the capability of describing the computation process by a program.  Tangent doesn't support all Python syntax used in the description of the forward pass. Similarly, we might not allow all Python syntax used to describe the workflow, if we follow the Tangent way.  The steps in Tagnent include some pre-listed functions, mostly, TensorFlow operations, and Python operators.
+
+## Concepts
+
+Some base jobs.
+
+WorkflowParam represents intermediate values passed between steps, and can also be used to find out dependencies. (name, owner_step, type).
+How inputs/outputs are supported in Argo now?
```
Comment (merlintang) [inline doc/design/design_workflow.md:40]: inputs and outputs can be pvolume, or other file storage place.

File: doc/design/design_workflow.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        success_condition: str,
+        failure_condition: str,
+        **kwargs,
+    ):
+        # Create a new instance of ResourceJob
+
+```
+
+**Branching**
+```python
```
Comment (merlintang) [inline doc/design/design_workflow.md:83]: confused here, branching point to the job to control this ? or the workflow or pipeline control?

File: doc/design/design_workflow.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+### Google Tangent
+
+Tangent is another deep learning system developed by Google. In contrast to TensorFlow, it represents the computation process by Python source code, other than a graph.
+
+Tangent does autodiff by parsing the Python source code into an abstract syntax tree (AST), adding the backward steps into the AST tree, and printing the AST tree into a new snippet of Python code.
+
+We like the capability of describing the computation process by a program.  Tangent doesn't support all Python syntax used in the description of the forward pass. Similarly, we might not allow all Python syntax used to describe the workflow, if we follow the Tangent way.  The steps in Tagnent include some pre-listed functions, mostly, TensorFlow operations, and Python operators.
+
+## Concepts
```
Comment (merlintang) [inline doc/design/design_workflow.md:35]: do we need to define the pipeline or dag at first ? then follow the jobs, and control logic.

**Review by merlintang — COMMENTED** [review]
I leave small comments here.

**Review by wangkuiyi — APPROVED** [review]
Let us merge the current work ASAP, so could we go on with this design.

**Pull request merged**